source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2255,6 +2255,209 @@ void COpenGLDriver::SAuxContext::setActiveDescriptorSet(uint32_t index, const CO
+            UPDATE_STATE(rasterParams.rasterizerDiscardEnable);
+        }
+        //if (STATE_NEQ(rasterParams.polygonOffsetEnable))
+        if (STATE_NEQ(rasterParams.lineWidth)) {
+            glLineWidth(nextState.rasterParams.lineWidth);
+            UPDATE_STATE(rasterParams.lineWidth);
+        }
+        if (STATE_NEQ(rasterParams.sampleShadingEnable)) {
+            disable_enable_fptr[nextState.rasterParams.sampleShadingEnable](GL_SAMPLE_SHADING);
+            UPDATE_STATE(rasterParams.sampleShadingEnable);
+        }
+        if (nextState.rasterParams.sampleShadingEnable && STATE_NEQ(rasterParams.minSampleShading)) {
+            COpenGLExtensionHandler::extGlMinSampleShading(nextState.rasterParams.minSampleShading);
+            UPDATE_STATE(rasterParams.minSampleShading);
+        }
+        if (STATE_NEQ(rasterParams.sampleMaskEnable)) {
+            disable_enable_fptr[nextState.rasterParams.sampleMaskEnable](GL_SAMPLE_MASK);
+            UPDATE_STATE(rasterParams.sampleMaskEnable);
+        }
+        if (nextState.rasterParams.sampleMaskEnable && STATE_NEQ(rasterParams.sampleMask[0])) {
+            COpenGLExtensionHandler::extGlSampleMaski(0u, nextState.rasterParams.sampleMask[0]);
+            UPDATE_STATE(rasterParams.sampleMask[0]);
+        }
+        if (nextState.rasterParams.sampleMaskEnable && STATE_NEQ(rasterParams.sampleMask[1])) {
+            COpenGLExtensionHandler::extGlSampleMaski(1u, nextState.rasterParams.sampleMask[1]);
+            UPDATE_STATE(rasterParams.sampleMask[1]);
+        }
+        if (STATE_NEQ(rasterParams.depthWriteEnable)) {
+            glDepthMask(nextState.rasterParams.depthWriteEnable);
+            UPDATE_STATE(rasterParams.depthWriteEnable);
+        }
+#undef STATE_NEQ
+#undef UPDATE_STATE
+    }
+    if (stateBits & GSB_VAO_AND_VERTEX_INPUT)
+    {
+    }
+    if (stateBits & GSB_DESCRIPTOR_SETS)
+    {
+        auto rebindBuffers = [this](GLenum target, const auto& currBufs, const auto& nextBufs)
+        {
+            auto bufBndDiffers = [](const auto& curr, const auto& next, uint32_t ix) {
+                return curr.buffers[ix] != next.buffers[ix] || curr.offsets[ix] != next.offsets[ix] || curr.sizes[ix] != next.sizes[ix];
+            };
+#define BUF_NEQ(ix) bufBndDiffers(currBufs, nextBufs, ix)
+
+            constexpr GLuint invalid_ix = static_cast<GLuint>(-1);
+            GLuint first = invalid_ix;
+            GLsizei count = 0u;
+            for (uint32_t i = 0u; i < 16u; ++i)
+            {
+                if (BUF_NEQ(i)) {
+                    if (first == invalid_ix)
+                        first = i;
+                    ++count;
+                }
+                else {
+                    if (count)
+                        COpenGLExtensionHandler::extGlBindBuffersRange(target, first, count, nextBufs.buffers+first, nextBufs.offsets+first, nextBufs.sizes+first);
```

Comment:
dont fuck around with this optimization, if I change every other buffer in a set of 32, you'll be calling multibind 16 times.

If you really need to optimize then try and trim the front and the back of the range

Reply:
try and find ONE first and ONE count for ONE call

Reply:
Actually storing and shadowing all the bindpoints might be a huge overhead, just store descriptor set pointers and multibind everything if desc-set changes